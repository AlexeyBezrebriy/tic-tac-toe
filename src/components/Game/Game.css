.Game-board {
  width: 200px;
}
