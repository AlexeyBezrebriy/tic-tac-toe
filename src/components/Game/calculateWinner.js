export default function calculateWinner(squares) {
  const lines = [
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 4, 8],
    [2, 4, 6],
    [1, 4, 7],
  ]
  let winner = ""

  for (let index = 0; index < squares.length; index++) {
    let [a, b, c] = lines[index]
    if (squares[a]) {
      let player = squares[a]
      if (
        squares[a] === player &&
        squares[b] === player &&
        squares[c] === player
      ) {
        winner = `winner is ${player}`
      }
    }
  }
  return winner
}
