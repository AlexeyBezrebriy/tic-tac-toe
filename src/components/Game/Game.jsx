import { useState } from "react"
import Board from "../Board/Board"
import calculateWinner from "./calculateWinner"
import "./Game.css"

function Game() {
  const [state, setState] = useState({
    history: [
      {
        squares: Array(9).fill(null),
        xIsNext: true,
      },
    ],
    move: 0,
  })

  function clickChange(i) {
    const history = state.history.slice(0, state.move + 1)
    const currentSquares = state.history[history.length - 1].squares
    const currentPlayer = state.history[history.length - 1].xIsNext
    const squares = currentSquares.slice()
    if (winner || squares[i]) {
      return
    }
    squares[i] = currentPlayer ? "X" : "O"

    setState({
      history: history.concat([{ squares: squares, xIsNext: !currentPlayer }]),
      move: history.length,
    })
  }

  function moveHistory(move) {
    setState({ ...state, move: move })
  }

  const showHistory = state.history.map((item, move) => {
    return (
      <li key={move}>
        <button onClick={() => moveHistory(move)}>
          {move ? "Go to step" : "Go to start"}
        </button>
      </li>
    )
  })

  const currentSquares = state.history[state.move].squares
  const currentMove = state.history[state.move].xIsNext

  const winner = calculateWinner(currentSquares)
  let status = ""
  if (winner) {
    status = winner
  } else {
    status = `Next player ${currentMove ? "X" : "O"}`
  }

  return (
    <div className="Game-board">
      <Board
        status={status}
        squares={currentSquares}
        onClick={(i) => clickChange(i)}
      />

      <ol>{showHistory}</ol>
    </div>
  )
}

export default Game
