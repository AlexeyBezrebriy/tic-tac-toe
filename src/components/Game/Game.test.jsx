import "@testing-library/jest-dom/extend-expect"
import { render, screen } from "@testing-library/react"
import renderer from "react-test-renderer"
import Game from "./Game"

describe("Game test", () => {
  test("should display Board with history", () => {
    render(<Game />)
    const btn = screen.getAllByRole("button")
    expect(btn.length).toBe(10)
  })
  test("Game is render correctly", () => {
    const domTree = renderer.create(<Game />).toJSON()
    expect(domTree).toMatchSnapshot()
  })
})
