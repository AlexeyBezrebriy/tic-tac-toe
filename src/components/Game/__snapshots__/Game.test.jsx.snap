// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game test Game is render correctly 1`] = `
<div
  className="Game-board"
>
  <h2
    className="Board-head"
  >
    Next player X
  </h2>
  <div
    className="Board-grid"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <ol>
    <li>
      <button
        onClick={[Function]}
      >
        Go to start
      </button>
    </li>
  </ol>
</div>
`;
